src/domain/datasets/services/src/use_cases/update_version_file_use_case_impl.rs

```diff
@@ -0,0 +1,227 @@
+        &self,
+        dataset_ref: &odf::DatasetRef,
+    ) -> Result<(FileVersion, odf::Multihash), InternalError> {
+        // TODO: Consider retractions / corrections
+        let query_res = self
+            .query_svc
+            .tail(dataset_ref, 0, 1, GetDataOptions::default())
+            .await
+            .int_err()?;
+
+        let Some(df) = query_res.df else {
+            return Ok((0, query_res.block_hash));
+        };
+
+        let last_version = df
+            .select_columns(&["version"])
+            .int_err()?
+            .collect_scalar::<datafusion::arrow::datatypes::Int32Type>()
+            .await
+            .int_err()?;
+
+        let last_version = FileVersion::try_from(last_version.unwrap_or(0)).unwrap();
+
+        Ok((last_version, query_res.block_hash))
+    }
+
+    async fn get_content_args_from_latest_entry(
+        &self,
+        dataset_ref: &odf::DatasetRef,
+    ) -> Result<Option<ContentArgs>, InternalError> {
+        // TODO: Consider retractions / corrections
+        let query_res = self
+            .query_svc
+            .tail(dataset_ref, 0, 1, GetDataOptions::default())
+            .await
+            .int_err()?;
+
+        let Some(df) = query_res.df else {
+            return Ok(None);
+        };
+
+        let records = df.collect_json_aos().await.int_err()?;
+
+        assert_eq!(records.len(), 1);
+        let Value::Object(mut record) = records.into_iter().next().unwrap() else {
+            unreachable!()
+        };
+
+        // TODO: Restrict after migration
+        let content_length = usize::try_from(
+            record
+                .remove("content_length")
+                .unwrap_or_default()
+                .as_u64()
+                .unwrap_or_default(),
+        )
+        .unwrap();
+        let content_type =
+            MediaType::from(record.remove("content_type").unwrap().as_str().unwrap());
+        let content_hash = odf::Multihash::from_multibase(
+            record.remove("content_hash").unwrap().as_str().unwrap(),
+        )
+        .unwrap();
+
+        Ok(Some(ContentArgs {
+            content_length,
+            content_stream: None,
+            content_hash,
+            content_type: Some(content_type),
+        }))
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[common_macros::method_names_consts]
+#[async_trait::async_trait]
+impl UpdateVersionFileUseCase for UpdateVersionFileUseCaseImpl {
+    #[tracing::instrument(level = "info", name = UpdateVersionFileUseCaseImpl_execute, skip_all, fields(%dataset_handle.id))]
+    async fn execute(
+        &self,
+        dataset_handle: &odf::DatasetHandle,
+        content_args_maybe: Option<ContentArgs>,
+        expected_head: Option<odf::Multihash>,
+        extra_data: Option<ExtraDataFields>,
+    ) -> Result<UpdateVersionFileResult, UpdateVersionFileUseCaseError> {
```

Comment:
Where is an authorization check?
If it is still in the API layer, consider moving it here (except for basic authentication).